### PR TITLE
elastic_scale: added filter for elastic scaling

### PR DESCRIPTION
**Pull request by schnoellm**
Hi there!

This elastic_scale filter allows to scale non-linearly scale video footage.
Combined with linear scaling one can scale 4:3 video footage to 16:9, as it is for example used with GoPro's Supverview (https://gopro.com/help/articles/question_answer/What-is-SuperView).

Feel free to give feedback on improvements.

BR

### CODE REVIEW

File: src/filter/elastic_scale/elastic_scale.cpp [hunk 1]
```diff
@@ -54,10 +54,10 @@ class ElasticScale : public frei0r::filter
...
 
     {
-        register_param(m_scaleCenter,"scaleCenter","Center position of the linearly scaled area");
-        register_param(m_linearScaleArea,"linearScaleArea","Area of the linearly scaled area");
-        register_param(m_linearScaleFactor,"linearScaleFactor","Amount how much the image is scaled");
-        register_param(m_nonLinearScaleFactor,"polynomialFactor","Polynomial factor for the transition between linear and non-linear scaling");
+        register_param(m_scaleCenter,"scaleCenter","Center position of the linear scaled area");
+        register_param(m_linearScaleArea,"linearScaleArea","Width of linear sclaed area");
+        register_param(m_linearScaleFactor,"linearScaleFactor","Amount how much the linearScaleArea is scaled");
+        register_param(m_nonLinearScaleFactor,"nonLinearScaleFactor","Amount how much the outer left and outer righter area are scaled non linearly");
```
Comment (ddennedy) [inline src/filter/elastic_scale/elastic_scale.cpp]: Please be aware that some applications use the name (scaleCenter, etc.) as the label in their generated UI. So, a non-identifier style name is more appropriate: Center, Linear Width, Linear Scale Factor, Non-linear Scale Factor

File: src/filter/elastic_scale/elastic_scale.cpp [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    f0r_param_double m_scaleCenter;
+    f0r_param_double m_linearScaleArea;
+    f0r_param_double m_linearScaleFactor;
+    f0r_param_double m_nonLinearScaleFactor;
+
+    ElasticScale(unsigned int width, unsigned int height) : m_width(width), m_height(height)
+
+    {
+        register_param(m_scaleCenter,"scaleCenter","Center position of the linear scaled area");
+        register_param(m_linearScaleArea,"linearScaleArea","Width of linear sclaed area");
```
Comment (ddennedy) [inline src/filter/elastic_scale/elastic_scale.cpp]: misspelling here "sclaed"

**Comment by ddennedy** [comment]
This filter is a great idea! However, I get some bizarre results when I try to experiment with linearScaleArea or linearScaleFactor not zero as well as nonLinearScaleFactor larger than, say, 0.3.
What further guidance can you give about using those?
Does the mapping from the [0,1] range of these parameters need tuning to make it more usable?

File: src/filter/elastic_scale/elastic_scale.cpp (hunk 1 above)
Comment (schnoellm) [reply src/filter/elastic_scale/elastic_scale.cpp]: Do you mean changing the respective internal member variables of the class, or just renaming the parameter names, those member variables are registered at by register_param?

**Comment by schnoellm** [comment]
What do you think about using a range of [-1,1] and map it internally to the respective limits of [-0.2,0.2]?
The reason why I prefer that range is because the filter would enable you to apply exactly 'zero' non-linearity (resp. linear scaling) with a value of 0, a certain sine based non-linearity with the value 1 and the exact opposit amount (also speaking of visible opposit) with a value of -1.
Also it makes more sense to me when the 'zero' non-linearity is with value 0 instead of 0.5.
What do you think about this approach?

**Comment by rrrapha** [comment]
It's not a free choice, it is how the interface is defined. ;)
Applications like kdenlive or shotcut will assume a range of [0,1] and display a
 slider using that range.

This is from frei0r.h:

```
/**
 * The double type. The allowed range of values is [0, 1].
 */
typedef double f0r_param_double;
```

**Comment by schnoellm** [comment]
Ok, that's a valid point ;-)
I will change that accordingly.

**Comment by ddennedy** [comment]
Obviously @rrrapha is correct, but also you do not need to strictly enforce the range in your plugin unless going outside would cause a crash.  Best practice is to map at least 90% of the most usable actual range into [0, 1] where going beyond gives undefined results, but some users may try to experiment, which is fine (assuming their interface permits it)!

**Comment by schnoellm** [comment]
Just changed the input ranges for all parameters to fit into [0,1]. If the input exceeds this range the value is adjusted automatically.
The only value currently mapped internally is nonLinearScaleFactor, which maps to [-0.2,0.2].
I think now it should be fine to be used in apps like kdenlive and shotcut later on.

File: src/filter/elastic_scale/elastic_scale.cpp [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    f0r_param_double m_prev_linearScaleFactor;
+    f0r_param_double m_prev_nonLinearScaleFactor;
+
+    f0r_param_double m_intern_scaleCenter;
+    f0r_param_double m_intern_linearScaleArea;
+    f0r_param_double m_intern_linearScaleFactor;
+    f0r_param_double m_intern_nonLinearScaleFactor;
+
+    unsigned int m_width;
+    unsigned int m_height;
```
[the remark this answers is not in the document]
Comment (schnoellm) [reply src/filter/elastic_scale/elastic_scale.cpp]: It is possible to use 'width' and 'height' throughout the whole class and its methods, but this involves setting the public members 'width' and 'height' for the base class 'fx' in the constructor 'ElasticScale'.
It seems like a bug, that 'width' and 'height' are actually correctly injected to the constructor but not set for the base class 'fx'. Without calling 'this->width = width; this->height = height' in the 'ElasticScale' constructor those are never set to reasonable values.

**Comment by schnoellm** [comment]
@rrrapha the crash due to certain parameters should be fixed with the latest commit. Please also share the input parameters so I can try to reproduce the situation.

**Comment by rrrapha** [comment]
@schnoellm yes the crash is now gone ;)

..but there seems to be other problems:
Can you try the following with an image size of 1290x849?
`ffmpeg -i in.jpg -vf "frei0r=elastic_scale:0.2|0.2|0.2|0.2" out.jpg`
The resulting image looks very funny here!
With ohter sizes like 2048x1330 I'm seeing strange pixels in the first column.

File: src/filter/elastic_scale/elastic_scale.cpp [hunk 4]
```diff
@@ -174,51 +174,57 @@ class ElasticScale : public frei0r::filter
...
-        if (m_borderXAbsLeftSrc <= 1)
+        // validate if resulting borders are still within the valid range for later calculations [1,width-1]
+        // negative results are limited
+        if (tmp_borderXAbsLeftSrc <= 1)
         { m_borderXAbsLeftSrc = 1; }
-        else if (m_borderXAbsLeftSrc >= m_width-1)
-        { m_borderXAbsLeftSrc = m_width-1; }
+        else if (tmp_borderXAbsLeftSrc >= (int)width-1)
+        { m_borderXAbsLeftSrc = width-1; }
+        else { m_borderXAbsLeftSrc = tmp_borderXAbsLeftSrc; }
```
Comment (rrrapha) [inline src/filter/elastic_scale/elastic_scale.cpp]: IMO these tmp_ variables are not needed. Wouldn't it make sense to just fix the compare like this:
`if ((int)m_borderXAbsLeftSrc <= 1)`

File: src/filter/elastic_scale/elastic_scale.cpp [hunk 5]
```diff
@@ -122,20 +118,25 @@ class ElasticScale : public frei0r::filter
     }
 
 private:
-    
+
+    // input params    
+    f0r_param_double m_scaleCenter;
+    f0r_param_double m_linearScaleArea;
+    f0r_param_double m_linearScaleFactor;
+    f0r_param_double m_nonLinearScaleFactor;
```
Comment (rrrapha) [inline src/filter/elastic_scale/elastic_scale.cpp]: Note that the double and bool param types have recently been changed in all plugins.
This does work for the double type, but using f0r_param_bool would be a mistake. Therefore f0r_param_double is misleading..

Comment (schnoellm) [reply src/filter/elastic_scale/elastic_scale.cpp]: Changed the data types from f0r_param_double to double

**Comment by rrrapha** [comment]
@schnoellm Thanks for the explanation, obviously width/height should be a multiple of 8. 
I think it doesn't hurt if other sizes work as well though ;)
The filter now works fine for me (doesn't crash and looks reasonable).

Maybe this is pedantic, but frei0r.h also has this comment:
```
 * This funcition should not alter the parameters of the effect in any
 * way (\ref f0r_get_param_value should return the same values after a call
 * to \ref f0r_update as before the call).
```
Your code does change the parameter values if they are outside the range [0,1].
I'd be curious how @ddennedy thinks about this.

**Comment by schnoellm** [comment]
@rrrapha You are right, it is indeed a little pedantic. To be honest, you could have also mentioned that thing with the parameters earlier ;-)
Nevertheless I've changed that in the code, so the external parameters are not modified and instead only the internal parameters are limited to [0,1] and mapped accordingly.

**Comment by rrrapha** [comment]
@schnoellm sorry, I didnt't notice this earlier.. 
It was meant as a question rather than an actual change request. (I can't speak for the project owners, I just thought the behaviour should match the API description)

**Comment by schnoellm** [comment]
@rrrapha it's ok ;-)
As it was only a minor change in the code, it wasn't a big deal. I aggree, if there is the possibility to adhere to api constraints, it's valid to point out to them.

**Pull request merged**